File: Java/src/main/java/OneLang/Parsers/Common/ExpressionParser/ExpressionParser.java
package OneLang.Parsers.Common.ExpressionParser;

import OneLang.Parsers.Common.Reader.Reader;
import OneLang.Parsers.Common.NodeManager.NodeManager;
import OneLang.One.Ast.AstTypes.UnresolvedType;
import OneLang.One.Ast.Expressions.Expression;
import OneLang.One.Ast.Expressions.MapLiteral;
import OneLang.One.Ast.Expressions.ArrayLiteral;
import OneLang.One.Ast.Expressions.UnaryType;
import OneLang.One.Ast.Expressions.Identifier;
import OneLang.One.Ast.Expressions.NumericLiteral;
import OneLang.One.Ast.Expressions.StringLiteral;
import OneLang.One.Ast.Expressions.UnresolvedCallExpression;
import OneLang.One.Ast.Expressions.CastExpression;
import OneLang.One.Ast.Expressions.BinaryExpression;
import OneLang.One.Ast.Expressions.UnaryExpression;
import OneLang.One.Ast.Expressions.ParenthesizedExpression;
import OneLang.One.Ast.Expressions.ConditionalExpression;
import OneLang.One.Ast.Expressions.ElementAccessExpression;
import OneLang.One.Ast.Expressions.PropertyAccessExpression;
import OneLang.One.Ast.Expressions.MapLiteralItem;
import OneLang.Utils.ArrayHelper.ArrayHelper;
import OneLang.One.Ast.Interfaces.IType;

import OneLang.Parsers.Common.ExpressionParser.Operator;
import java.util.Map;
import OneLang.One.Ast.Interfaces.IType;
import OneLang.Parsers.Common.Reader.Reader;
import OneLang.Parsers.Common.ExpressionParser.IExpressionParserHooks;
import OneLang.Parsers.Common.NodeManager.NodeManager;
import OneLang.Parsers.Common.ExpressionParser.ExpressionParserConfig;
import OneLang.Parsers.Common.ExpressionParser.PrecedenceLevel;
import OneStd.Objects;
import java.util.Arrays;
import java.util.LinkedHashMap;
import OneLang.One.Ast.Expressions.MapLiteral;
import OneLang.One.Ast.Expressions.MapLiteralItem;
import java.util.ArrayList;
import OneLang.One.Ast.Expressions.ArrayLiteral;
import OneLang.One.Ast.Expressions.Expression;
import OneLang.One.Ast.Expressions.UnaryExpression;
import OneLang.One.Ast.Expressions.Identifier;
import OneLang.One.Ast.Expressions.NumericLiteral;
import OneLang.One.Ast.Expressions.StringLiteral;
import OneLang.One.Ast.Expressions.ParenthesizedExpression;
import OneLang.One.Ast.Expressions.BinaryExpression;
import OneLang.One.Ast.Expressions.ConditionalExpression;
import OneLang.One.Ast.Expressions.UnresolvedCallExpression;
import OneLang.One.Ast.Expressions.ElementAccessExpression;
import OneLang.One.Ast.Expressions.PropertyAccessExpression;
import OneLang.One.Ast.Expressions.CastExpression;
import OneLang.One.Ast.AstTypes.UnresolvedType;

public class ExpressionParser {
    public Map<String, Operator> operatorMap;
    public String[] operators;
    public Integer prefixPrecedence;
    public IType stringLiteralType;
    public IType numericLiteralType;
    public Reader reader;
    public IExpressionParserHooks hooks;
    public NodeManager nodeManager;
    public ExpressionParserConfig config;
    
    public ExpressionParser(Reader reader, IExpressionParserHooks hooks, NodeManager nodeManager, ExpressionParserConfig config)
    {
        this.reader = reader;
        this.hooks = hooks;
        this.nodeManager = nodeManager;
        this.config = config;
        this.stringLiteralType = null;
        this.numericLiteralType = null;
        if (this.config == null)
            this.config = ExpressionParser.defaultConfig();
        this.reconfigure();
    }
    
    public static ExpressionParserConfig defaultConfig() {
        var config = new ExpressionParserConfig();
        config.unary = new String[] { "++", "--", "!", "not", "+", "-", "~" };
        config.precedenceLevels = new PrecedenceLevel[] { new PrecedenceLevel("assignment", new String[] { "=", "+=", "-=", "*=", "/=", "<<=", ">>=" }, true), new PrecedenceLevel("conditional", new String[] { "?" }, false), new PrecedenceLevel("or", new String[] { "||", "or" }, true), new PrecedenceLevel("and", new String[] { "&&", "and" }, true), new PrecedenceLevel("comparison", new String[] { ">=", "!=", "===", "!==", "==", "<=", ">", "<" }, true), new PrecedenceLevel("sum", new String[] { "+", "-" }, true), new PrecedenceLevel("product", new String[] { "*", "/", "%" }, true), new PrecedenceLevel("bitwise", new String[] { "|", "&", "^" }, true), new PrecedenceLevel("exponent", new String[] { "**" }, true), new PrecedenceLevel("shift", new String[] { "<<", ">>" }, true), new PrecedenceLevel("range", new String[] { "..." }, true), new PrecedenceLevel("in", new String[] { "in" }, true), new PrecedenceLevel("prefix", new String[0], false), new PrecedenceLevel("postfix", new String[] { "++", "--" }, false), new PrecedenceLevel("call", new String[] { "(" }, false), new PrecedenceLevel("propertyAccess", new String[0], false), new PrecedenceLevel("elementAccess", new String[] { "[" }, false) };
        config.rightAssoc = new String[] { "**" };
        config.aliases = Map.of("===", "==", "!==", "!=", "not", "!", "and", "&&", "or", "||");
        config.propertyAccessOps = new String[] { ".", "::" };
        return config;
    }
    
    public void reconfigure() {
        Arrays.stream(this.config.precedenceLevels).filter(x -> Objects.equals(x.name, "propertyAccess")).findFirst().orElse(null).operators = this.config.propertyAccessOps;
        
        this.operatorMap = new LinkedHashMap<String, Operator>();
        
        for (Integer i = 0; i < this.config.precedenceLevels.length; i++) {
            var level = this.config.precedenceLevels[i];
            var precedence = i + 1;
            if (Objects.equals(level.name, "prefix"))
                this.prefixPrecedence = precedence;
            
            if (level.operators == null)
                continue;
            
            for (var opText : level.operators) {
                var op = new Operator(opText, precedence, level.binary, Arrays.stream(this.config.rightAssoc).anyMatch(opText::equals), Objects.equals(level.name, "postfix"));
                
                this.operatorMap.put(opText, op);
            }
        }
        
        this.operators = ArrayHelper.sortBy(this.operatorMap.keySet().toArray(String[]::new), x -> -x.length());
    }
    
    public MapLiteral parseMapLiteral(String keySeparator, String startToken, String endToken) {
        if (!this.reader.readToken(startToken))
            return null;
        
        var items = new ArrayList<MapLiteralItem>();
        do {
            if (this.reader.peekToken(endToken))
                break;
            
            var name = this.reader.readString();
            if (name == null)
                name = this.reader.expectIdentifier("expected string or identifier as map key");
            
            this.reader.expectToken(keySeparator, null);
            var initializer = this.parse(0, true);
            items.add(new MapLiteralItem(name, initializer));
        } while (this.reader.readToken(","));
        
        this.reader.expectToken(endToken, null);
        return new MapLiteral(items.toArray(MapLiteralItem[]::new));
    }
    
    public ArrayLiteral parseArrayLiteral(String startToken, String endToken) {
        if (!this.reader.readToken(startToken))
            return null;
        
        var items = new ArrayList<Expression>();
        if (!this.reader.readToken(endToken)) {
            do {
                var item = this.parse(0, true);
                items.add(item);
            } while (this.reader.readToken(","));
            
            this.reader.expectToken(endToken, null);
        }
        return new ArrayLiteral(items.toArray(Expression[]::new));
    }
    
    public Expression parseLeft(Boolean required) {
        var result = this.hooks != null ? this.hooks.unaryPrehook() : null;
        if (result != null)
            return result;
        
        var unary = this.reader.readAnyOf(this.config.unary);
        if (unary != null) {
            var right = this.parse(this.prefixPrecedence, true);
            return new UnaryExpression(UnaryType.Prefix, unary, right);
        }
        
        var id = this.reader.readIdentifier();
        if (id != null)
            return new Identifier(id);
        
        var num = this.reader.readNumber();
        if (num != null)
            return new NumericLiteral(num);
        
        var str = this.reader.readString();
        if (str != null)
            return new StringLiteral(str);
        
        if (this.reader.readToken("(")) {
            var expr = this.parse(0, true);
            this.reader.expectToken(")", null);
            return new ParenthesizedExpression(expr);
        }
        
        if (required)
            this.reader.fail("unknown (literal / unary) token in expression", -1);
        
        return null;
    }
    
    public Operator parseOperator() {
        for (var opText : this.operators) {
            if (this.reader.peekToken(opText))
                return this.operatorMap.get(opText);
        }
        
        return null;
    }
    
    public Expression[] parseCallArguments() {
        var args = new ArrayList<Expression>();
        
        if (!this.reader.readToken(")")) {
            do {
                var arg = this.parse(0, true);
                args.add(arg);
            } while (this.reader.readToken(","));
            
            this.reader.expectToken(")", null);
        }
        
        return args.toArray(Expression[]::new);
    }
    
    public void addNode(Object node, Integer start) {
        if (this.nodeManager != null)
            this.nodeManager.addNode(node, start);
    }
    
    public Expression parse(Integer precedence, Boolean required) {
        this.reader.skipWhitespace(false);
        var leftStart = this.reader.offset;
        var left = this.parseLeft(required);
        if (left == null)
            return null;
        this.addNode(left, leftStart);
        
        while (true) {
            if (this.hooks != null) {
                var parsed = this.hooks.infixPrehook(left);
                if (parsed != null) {
                    left = parsed;
                    this.addNode(left, leftStart);
                    continue;
                }
            }
            
            var op = this.parseOperator();
            if (op == null || op.precedence <= precedence)
                break;
            this.reader.expectToken(op.text, null);
            var opText = this.config.aliases.containsKey(op.text) ? this.config.aliases.get(op.text) : op.text;
            
            if (op.isBinary) {
                var right = this.parse(op.isRightAssoc ? op.precedence - 1 : op.precedence, true);
                left = new BinaryExpression(left, opText, right);
            }
            else if (op.isPostfix)
                left = new UnaryExpression(UnaryType.Postfix, opText, left);
            else if (Objects.equals(op.text, "?")) {
                var whenTrue = this.parse(0, true);
                this.reader.expectToken(":", null);
                var whenFalse = this.parse(op.precedence - 1, true);
                left = new ConditionalExpression(left, whenTrue, whenFalse);
            }
            else if (Objects.equals(op.text, "(")) {
                var args = this.parseCallArguments();
                left = new UnresolvedCallExpression(left, new IType[0], args);
            }
            else if (Objects.equals(op.text, "[")) {
                var elementExpr = this.parse(0, true);
                this.reader.expectToken("]", null);
                left = new ElementAccessExpression(left, elementExpr);
            }
            else if (Arrays.stream(this.config.propertyAccessOps).anyMatch(op.text::equals)) {
                var prop = this.reader.expectIdentifier("expected identifier as property name");
                left = new PropertyAccessExpression(left, prop);
            }
            else
                this.reader.fail("parsing '" + op.text + "' is not yet implemented", -1);
            
            this.addNode(left, leftStart);
        }
        
        if (left instanceof ParenthesizedExpression && ((ParenthesizedExpression)left).expression instanceof Identifier) {
            var expr = this.parse(0, false);
            if (expr != null)
                return new CastExpression(new UnresolvedType(((Identifier)((ParenthesizedExpression)left).expression).text, new IType[0]), expr);
        }
        
        return left;
    }
}

File: Python/OneLang/Parsers/Common/ExpressionParser.py
from OneLangStdLib import *
import OneLang.Parsers.Common.Reader as read
import OneLang.Parsers.Common.NodeManager as nodeMan
import OneLang.One.Ast.AstTypes as astTypes
import OneLang.One.Ast.Expressions as exprs
import OneLang.Utils.ArrayHelper as arrayHelp
import OneLang.One.Ast.Interfaces as ints

class Operator:
    def __init__(self, text, precedence, is_binary, is_right_assoc, is_postfix):
        self.text = text
        self.precedence = precedence
        self.is_binary = is_binary
        self.is_right_assoc = is_right_assoc
        self.is_postfix = is_postfix

class PrecedenceLevel:
    def __init__(self, name, operators, binary):
        self.name = name
        self.operators = operators
        self.binary = binary

class ExpressionParserConfig:
    def __init__(self):
        self.unary = None
        self.precedence_levels = None
        self.right_assoc = None
        self.aliases = None
        self.property_access_ops = None

class ExpressionParser:
    def __init__(self, reader, hooks = None, node_manager = None, config = None):
        self.operator_map = None
        self.operators = None
        self.prefix_precedence = None
        self.string_literal_type = None
        self.numeric_literal_type = None
        self.reader = reader
        self.hooks = hooks
        self.node_manager = node_manager
        self.config = config
        if self.config == None:
            self.config = ExpressionParser.default_config()
        self.reconfigure()
    
    @classmethod
    def default_config(cls):
        config = ExpressionParserConfig()
        config.unary = ["++", "--", "!", "not", "+", "-", "~"]
        config.precedence_levels = [PrecedenceLevel("assignment", ["=", "+=", "-=", "*=", "/=", "<<=", ">>="], True), PrecedenceLevel("conditional", ["?"], False), PrecedenceLevel("or", ["||", "or"], True), PrecedenceLevel("and", ["&&", "and"], True), PrecedenceLevel("comparison", [">=", "!=", "===", "!==", "==", "<=", ">", "<"], True), PrecedenceLevel("sum", ["+", "-"], True), PrecedenceLevel("product", ["*", "/", "%"], True), PrecedenceLevel("bitwise", ["|", "&", "^"], True), PrecedenceLevel("exponent", ["**"], True), PrecedenceLevel("shift", ["<<", ">>"], True), PrecedenceLevel("range", ["..."], True), PrecedenceLevel("in", ["in"], True), PrecedenceLevel("prefix", [], False), PrecedenceLevel("postfix", ["++", "--"], False), PrecedenceLevel("call", ["("], False), PrecedenceLevel("propertyAccess", [], False), PrecedenceLevel("elementAccess", ["["], False)]
        config.right_assoc = ["**"]
        config.aliases = {
            "===": "==",
            "!==": "!=",
            "not": "!",
            "and": "&&",
            "or": "||"
        }
        config.property_access_ops = [".", "::"]
        return config
    
    def reconfigure(self):
        next(filter(lambda x: x.name == "propertyAccess", self.config.precedence_levels), None).operators = self.config.property_access_ops
        
        self.operator_map = {}
        
        i = 0
        
        while i < len(self.config.precedence_levels):
            level = self.config.precedence_levels[i]
            precedence = i + 1
            if level.name == "prefix":
                self.prefix_precedence = precedence
            
            if level.operators == None:
                continue
            
            for op_text in level.operators:
                op = Operator(op_text, precedence, level.binary, op_text in self.config.right_assoc, level.name == "postfix")
                
                self.operator_map[op_text] = op
            i = i + 1
        
        self.operators = sorted(self.operator_map.keys(), key=lambda x: -len(x))
    
    def parse_map_literal(self, key_separator = ":", start_token = "{", end_token = "}"):
        if not self.reader.read_token(start_token):
            return None
        
        items = []
        while True:
            if self.reader.peek_token(end_token):
                break
            
            name = self.reader.read_string()
            if name == None:
                name = self.reader.expect_identifier("expected string or identifier as map key")
            
            self.reader.expect_token(key_separator)
            initializer = self.parse()
            items.append(exprs.MapLiteralItem(name, initializer))
            if not (self.reader.read_token(",")):
                break
        
        self.reader.expect_token(end_token)
        return exprs.MapLiteral(items)
    
    def parse_array_literal(self, start_token = "[", end_token = "]"):
        if not self.reader.read_token(start_token):
            return None
        
        items = []
        if not self.reader.read_token(end_token):
            while True:
                item = self.parse()
                items.append(item)
                if not (self.reader.read_token(",")):
                    break
            
            self.reader.expect_token(end_token)
        return exprs.ArrayLiteral(items)
    
    def parse_left(self, required = True):
        result = self.hooks.unary_prehook() if self.hooks != None else None
        if result != None:
            return result
        
        unary = self.reader.read_any_of(self.config.unary)
        if unary != None:
            right = self.parse(self.prefix_precedence)
            return exprs.UnaryExpression(exprs.UNARY_TYPE.PREFIX, unary, right)
        
        id = self.reader.read_identifier()
        if id != None:
            return exprs.Identifier(id)
        
        num = self.reader.read_number()
        if num != None:
            return exprs.NumericLiteral(num)
        
        str = self.reader.read_string()
        if str != None:
            return exprs.StringLiteral(str)
        
        if self.reader.read_token("("):
            expr = self.parse()
            self.reader.expect_token(")")
            return exprs.ParenthesizedExpression(expr)
        
        if required:
            self.reader.fail(f'''unknown (literal / unary) token in expression''')
        
        return None
    
    def parse_operator(self):
        for op_text in self.operators:
            if self.reader.peek_token(op_text):
                return self.operator_map.get(op_text)
        
        return None
    
    def parse_call_arguments(self):
        args = []
        
        if not self.reader.read_token(")"):
            while True:
                arg = self.parse()
                args.append(arg)
                if not (self.reader.read_token(",")):
                    break
            
            self.reader.expect_token(")")
        
        return args
    
    def add_node(self, node, start):
        if self.node_manager != None:
            self.node_manager.add_node(node, start)
    
    def parse(self, precedence = 0, required = True):
        self.reader.skip_whitespace()
        left_start = self.reader.offset
        left = self.parse_left(required)
        if left == None:
            return None
        self.add_node(left, left_start)
        
        while True:
            if self.hooks != None:
                parsed = self.hooks.infix_prehook(left)
                if parsed != None:
                    left = parsed
                    self.add_node(left, left_start)
                    continue
            
            op = self.parse_operator()
            if op == None or op.precedence <= precedence:
                break
            self.reader.expect_token(op.text)
            op_text = self.config.aliases.get(op.text) if op.text in self.config.aliases else op.text
            
            if op.is_binary:
                right = self.parse(op.precedence - 1 if op.is_right_assoc else op.precedence)
                left = exprs.BinaryExpression(left, op_text, right)
            elif op.is_postfix:
                left = exprs.UnaryExpression(exprs.UNARY_TYPE.POSTFIX, op_text, left)
            elif op.text == "?":
                when_true = self.parse()
                self.reader.expect_token(":")
                when_false = self.parse(op.precedence - 1)
                left = exprs.ConditionalExpression(left, when_true, when_false)
            elif op.text == "(":
                args = self.parse_call_arguments()
                left = exprs.UnresolvedCallExpression(left, [], args)
            elif op.text == "[":
                element_expr = self.parse()
                self.reader.expect_token("]")
                left = exprs.ElementAccessExpression(left, element_expr)
            elif op.text in self.config.property_access_ops:
                prop = self.reader.expect_identifier("expected identifier as property name")
                left = exprs.PropertyAccessExpression(left, prop)
            else:
                self.reader.fail(f'''parsing \'{op.text}\' is not yet implemented''')
            
            self.add_node(left, left_start)
        
        if isinstance(left, exprs.ParenthesizedExpression) and isinstance(left.expression, exprs.Identifier):
            expr = self.parse(0, False)
            if expr != None:
                return exprs.CastExpression(astTypes.UnresolvedType(left.expression.text, []), expr)
        
        return left